### PR TITLE
Add Buf cli for updating repo settings

### PR DESCRIPTION
**Pull request by rubensf**

### CODE REVIEW

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.Visibility,
+		visibilityFlagName,
+		"",
+		fmt.Sprintf(`The repository's visibility setting. Must be one of %s.`, stringutil.SliceToString(allVisibiltyStrings)),
+	)
```
Comment (cyinma) [inline private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:75]: we might want to move this binding and the above const and variables to `private/buf/bufcli/bufcli.go` and share it with repository create, you can refer to e.g. `BindDisableSymlinks`

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+func visibilityFlagToVisibility(visibility string) (registryv1alpha1.Visibility, error) {
+	switch visibility {
+	case publicVisibility:
+		return registryv1alpha1.Visibility_VISIBILITY_PUBLIC, nil
+	case privateVisibility:
+		return registryv1alpha1.Visibility_VISIBILITY_PRIVATE, nil
+	default:
+		return 0, fmt.Errorf("invalid visibility: %s, expected one of %s", visibility, stringutil.SliceToString(allVisibiltyStrings))
+	}
+}
```
Comment (cyinma) [inline private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:129]: this conversion method too

**Comment by rubensf** [comment]
Done

File: private/buf/cmd/buf/command/beta/registry/repository/repositorycreate/repositorycreate.go [hunk 3]
```diff
@@ -73,18 +70,7 @@ func newFlags() *flags {
 }
 
 func (f *flags) Bind(flagSet *pflag.FlagSet) {
-	flagSet.StringVar(
-		&f.Format,
-		formatFlagName,
-		bufprint.FormatText.String(),
-		fmt.Sprintf(`The output format to use. Must be one of %s.`, bufprint.AllFormatsString),
-	)
```
Comment (robbertvanginkel) [inline private/buf/cmd/buf/command/beta/registry/repository/repositorycreate/repositorycreate.go:81]: Why did the format flag go away here?

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+type flags struct {
+	Visibility string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	bufcli.BindVisibility(flagSet, &f.Visibility, visibilityFlagName)
```
Comment (robbertvanginkel) [inline private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:68]: Making this a generic update command with a visibility flag seems to suggest that the plan is to add more flags here. However, the implementation calls a visibility specific update endpoint.

This doesn't have to be problematic, but if we add new flags here if each flag needs to issue it's own rpc then the cli interface on errors gets a bit fiddly. What if one of the rpc's succeeds but others fail? You'd likely want to fail, but if this command fails would you expect that maybe some of the updates did succeed?

I would expect that a mutating cli command with multiple different fields to update does so atomically, and that we should either move this to a `update-visibility` command or have a plan for a more generic update endpoint. cc @amckinney wondering if there's some precedent in the cli for this?

Comment (rubensf) [reply private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:68]: IIUC this is intended as one-rpc-addresses all, see [the design doc](https://www.notion.so/bufbuild/Resource-s-visibility-update-fae6fd07d9e74dc18b5cf1f5b38173d3) and the service call itself which [suggests multiple settings](https://github.com/bufbuild/buf/blob/main/proto/buf/alpha/registry/v1alpha1/repository.proto#L105). With the caveat that adding options means updating the RPC call...

We could instead make `UpdateRepositorySettingsByNameRequest` take a `UpdateRepositorySettingsByNameOptions`, which is really just a semantic difference that makes future changes non-breaking. Wdyt?

File: private/buf/cmd/buf/command/beta/registry/repository/repositorycreate/repositorycreate.go (hunk 3 above)
Comment (rubensf) [reply private/buf/cmd/buf/command/beta/registry/repository/repositorycreate/repositorycreate.go]: Accident :D Fixed. Thanks for catching!

File: private/buf/bufcli/bufcli.go [hunk 5]
```diff
@@ -249,6 +254,20 @@ Symlinks are never followed in Windows.`,
 	)
 }
 
+var allVisibiltyStrings = []string{
+	publicVisibility,
+	privateVisibility,
+}
```
Comment (robbertvanginkel) [inline private/buf/bufcli/bufcli.go:260]: Nit: order globals towards top of the file, see https://www.notion.so/bufbuild/Go-Code-Standards-ff40382e6b104c1097a017f9565d54db#c6aeeed92adf40e189d7323ba0ccd11d

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	visibilityFlagName = "visibility"
+
+	publicVisibility  = "public"
+	privateVisibility = "private"
+)
+
+var allVisibiltyStrings = []string{
+	publicVisibility,
+	privateVisibility,
+}
```
Comment (robbertvanginkel) [inline private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:40]:
```suggestion
)
```

**Review by robbertvanginkel — APPROVED** [review]

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go (hunk 4 above)
Comment (robbertvanginkel) [reply private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:68]: As long as we're doing 1 rpc call this makes sense, I misread in the beginning. Resolving.

File: private/buf/bufcli/bufcli.go [hunk 7]
```diff
@@ -146,6 +151,11 @@ var (
 	// These digests are used to make sure that the data written is actually what we expect, and if it is not,
 	// we clear an entry from the cache, i.e. delete the relevant data directory.
 	v1CacheModuleSumRelDirPath = normalpath.Join("v1", "module", "sum")
+
+	allVisibiltyStrings = []string{
```
Comment (cyinma) [inline private/buf/bufcli/bufcli.go:155]: lets add a comment for it too

File: private/buf/bufcli/bufcli.go [hunk 8]
```diff
@@ -249,6 +259,15 @@ Symlinks are never followed in Windows.`,
 	)
 }
 
+func BindVisibility(flagSet *pflag.FlagSet, addr *string, flagName string) {
```
Comment (cyinma) [inline private/buf/bufcli/bufcli.go:262]: exported function need a comment

File: private/buf/bufcli/bufcli.go [hunk 9]
```diff
@@ -952,3 +971,15 @@ func parseFullyQualifiedPath(
 	}
 	return moduleReference.String(), components[1], nil
 }
+
+// VisibilityFlagToVisibility parses the given string as a registryv1alpha1.Visibility.
+func VisibilityFlagToVisibility(visibility string) (registryv1alpha1.Visibility, error) {
```
Comment (cyinma) [inline private/buf/bufcli/bufcli.go:976]: exported function goes above unexported functions

**Review by cyinma — COMMENTED** [review]
looking good now! just some comments and order stuff

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+type flags struct {
+	Visibility string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	bufcli.BindVisibility(flagSet, &f.Visibility, visibilityFlagName)
```
Comment (cyinma) [inline private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:60]: I know that we are expecting to do partial update in future, but should we also `MarkFlagRequired` for the visibility for now?

**Review by cyinma — COMMENTED** [review]
looking good! lets also add a change log as this added a new command

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go (hunk 10 above)
Comment (rubensf) [reply private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:60]: I don't think it's worth the trouble, particularly as it's a developer tool - which should likely understand that a change without options means no change :)

File: private/buf/bufcli/bufcli.go [hunk 11]
```diff
@@ -839,6 +860,18 @@ func ValidateErrorFormatFlagLint(errorFormatString string, errorFormatFlagName s
 	return validateErrorFormatFlag(buflint.AllFormatStrings, errorFormatString, errorFormatFlagName)
 }
 
+// VisibilityFlagToVisibility parses the given string as a registryv1alpha1.Visibility.
+func VisibilityFlagToVisibility(visibility string) (registryv1alpha1.Visibility, error) {
```
[the remark this answers is not in the document]
Comment (rubensf) [reply private/buf/bufcli/bufcli.go:854]: ah, great point. Mybad.

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go (hunk 10 above)
Comment (rubensf) [reply private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:60]: Ah, I think that's a bug. We should be able to accept unspecified visibilities (particularly in the future). Fixed

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+	bufcli.WarnBetaCommand(ctx, container)
+	moduleIdentity, err := bufmoduleref.ModuleIdentityForString(container.Arg(0))
+	if err != nil {
+		return appcmd.NewInvalidArgumentError(err.Error())
+	}
+	visibility, err := bufcli.VisibilityFlagToVisibilityAllowUnspecified(flags.Visibility)
+	if err != nil {
+		return appcmd.NewInvalidArgumentError(err.Error())
+	}
+
```
Comment (cyinma) [inline private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:77]: I think the blank line here is unnecessary per the code standard

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	apiProvider, err := bufcli.NewRegistryProvider(ctx, container)
+	if err != nil {
+		return err
+	}
+	service, err := apiProvider.NewRepositoryService(ctx, moduleIdentity.Remote())
+	if err != nil {
+		return err
+	}
+
```
Comment (cyinma) [inline private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go:86]: same here

File: CHANGELOG.md [hunk 14]
```diff
@@ -2,7 +2,8 @@
 
 ## [Unreleased]
 
-- No changes yet.
+- Add `buf beta registry repository update` command. So far only with support
+  to update visibility.
```
Comment (cyinma) [inline CHANGELOG.md:6]: will this be better?
```suggestion
- Add `buf beta registry repository update` command which support visibility
  update for now.
```

File: private/buf/bufcli/bufcli.go [hunk 15]
```diff
@@ -829,6 +850,32 @@ func ParseSourceAndType(
...
+	case publicVisibility:
+		return registryv1alpha1.Visibility_VISIBILITY_PUBLIC, nil
+	case privateVisibility:
+		return registryv1alpha1.Visibility_VISIBILITY_PRIVATE, nil
+	default:
+		return 0, fmt.Errorf("invalid visibility: %s, expected one of %s", visibility, stringutil.SliceToString(allVisibiltyStrings))
+	}
+}
+
+// VisibilityFlagToVisibilityAllowUnspecified parses the given string as a registryv1alpha1.Visibility.
```
Comment (cyinma) [inline private/buf/bufcli/bufcli.go:865]:
```suggestion
// VisibilityFlagToVisibilityAllowUnspecified parses the given string as a registryv1alpha1.Visibility, where an empty string will be parsed as unspecified.
```

**Review by cyinma — APPROVED** [review]
LGTM except the last few nits!

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go (hunk 12 above)
Comment (rubensf) [reply private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go]: cleaned

File: private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go (hunk 13 above)
Comment (rubensf) [reply private/buf/cmd/buf/command/beta/registry/repository/repositoryupdate/repositoryupdate.go]: cleaned

File: CHANGELOG.md (hunk 14 above)
Comment (rubensf) [reply CHANGELOG.md]: done

File: private/buf/bufcli/bufcli.go (hunk 15 above)
Comment (rubensf) [reply private/buf/bufcli/bufcli.go]: done

**Review by rubensf — COMMENTED** [review]
all done

**Pull request merged**